radixdlt-java/radixdlt-java/src/test/java/com/radixdlt/client/lib/api/sync/SyncRadixApiCreationTest.java

```diff
@@ -149,14 +160,11 @@ public void testBuildAndSubmitTransactionWithMessage() {
 				.onFailure(failure -> fail(failure.toString()))
 				.onSuccess(builtTransactionDTO -> assertEquals(UInt256.from(100000000000000000L), builtTransactionDTO.getFee()))
 				.map(builtTransactionDTO -> builtTransactionDTO.toFinalized(KEY_PAIR1))
-				.onSuccess(finalizedTransaction -> client.transaction().finalize(finalizedTransaction)
+				.onSuccess(finalizedTransaction -> client.transaction().finalize(finalizedTransaction, false)
 					.onSuccess(txDTO -> assertNotNull(txDTO.getTxId()))
-					.map(txDTO -> finalizedTransaction.withTxId(txDTO.getTxId()))
 					.onSuccess(submittableTransaction -> client.transaction().submit(submittableTransaction)
 						.onFailure(failure -> fail(failure.toString()))
-						.onSuccess(txDTO -> submittableTransaction.rawTxId()
-							.ifPresentOrElse(aid -> assertEquals(aid, txDTO.getTxId()), () -> fail("Should not happen")))))
-			);
+						.onSuccess(txDTO -> assertEquals(submittableTransaction.getTxId(), txDTO.getTxId())))));
```

Comment:
Seems like this is turning into callback hell? I question how hard it would be to debug if an exception were to occur?

Reply:
The client code never throws an exception (this is the exact purpose of using `Result<T>` as a value container). Any exception which may occur can be only in the user code, so it will be straightforward to nail it down.
Also, given the complexity of the code (three steps of transaction submission, where each step may fail + assertions necessary only for test), it is rather clean and far from being callback hell. In real use cases this code could be split into few smaller functions, like, for example, this is done in `SyncRadixApiTest` (`makeStake()` + `makeStake(RadixApi, UInt256)`, for example).